### ISSUE TITLE
Cant retrieve connection from the pool: Timed out in bb8

### ISSUE THREAD

**Issue by IceMan1910** [urbica/martin#496]
Hi all. Just started on this mapping tools quest :)
I'm not being able to solve what seems to be a simple issue with martin.

I have my postgres bd running (OSM is the bd name), it is already fed with data coming from osm2pgsql with flex output. No issues there. It has a points, lines, polygons table with geometries.
Now i want to serve my vector tiles with martin :)

I'm on windows 10, my postegres is running on a local server on ubuntu.
**Running on terminal:**
`PS C:\Users\[user]\Desktop\martin> .\martin.exe -c .\config.yaml`

**On yaml file I have:**
```
connection_string: "postgres://postgres:[pass]@[local_ip_for_ubuntu]:5432/OSM"
default_srid: 4326
listen_addresses: "0.0.0.0:4000"
```

I've tried without the config file, using connection string on the terminal. Same output:
```
[2022-11-29T16:13:46Z INFO  martin] Starting Martin v0.6.1
[2022-11-29T16:13:46Z INFO  martin] Using .\config.yaml
[2022-11-29T16:13:46Z INFO  martin::pg::db] Connecting to database
[2022-11-29T16:14:16Z ERROR martin] Can't retrieve connection from the pool: Timed out in bb8
```
I can connect just fine with pgadmin to the bd, and I ran the osm2pgsql process on this windows machine with output going to this same bd.
I can´t seem to make this work. What am I missing?
Thanks in advance.

**Comment by nyurik** [comment]
@IceMan1910 are you able to run `psql "postgres://postgres:[pass]@[local_ip_for_ubuntu]:5432/OSM"` ?  (assuming you have psql installed on your windows box)

**Comment by IceMan1910** [comment]
@nyurik I don't have it instaled locally. Does it make sense to install it, just to test?
With psql on the pgAdmin4 on this machine, I can connect fine.

I also have another ubuntu server on my local network. Should I install martin on ubuntu instead of windows 10?

**Comment by nyurik** [comment]
I think it would be helpful - it seems to be happening on your machine, plus having psql when working with PG is highly useful for any kind of simple querying

**Comment by nyurik** [comment]
P.S. To be honest, I would never use Windows for any kind of server-related activities :)

**Comment by IceMan1910** [comment]
@nyurik OK so... I'm really stuck here.
Steps done:
- On my other linux box, same config file as I had on windows.

`./martin --config config.yaml`

Same old, same old :\

```
[2022-12-02T04:01:24Z INFO  martin] Starting Martin v0.6.1
[2022-12-02T04:01:24Z INFO  martin] Using config.yaml
[2022-12-02T04:01:24Z INFO  martin::pg::db] Connecting to database
[2022-12-02T04:01:54Z ERROR martin] Can't retrieve connection from the pool: Timed out in bb8
```

Now, with psql client installed, and copying the connection string from the config file to make sure it is exactly the same:

```
[user]@[server]:~/martin$ psql "postgres://postgres:[pass]@[ip]:5432/OSM"
psql (14.5 (Ubuntu 14.5-0ubuntu0.22.04.1), server 15.1 (Ubuntu 15.1-1.pgdg22.04+1))
WARNING: psql major version 14, server major version 15.
         Some psql features might not work.
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

OSM=# select version();
                                                              version                                                              
-----------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 15.1 (Ubuntu 15.1-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, 64-bit
(1 row)

OSM=# 
```



It connects fine.
Only with martin it does not. 

What can it be?
Could it be something with Rust? I have nothing related to Rust installed on this machine.

**Comment by nyurik** [comment]
Thanks for testing, this is indeed weird.  Could you try this:
* try running martin without config, i.e. `./martin "postgres://postgres:[pass]@[ip]:5432/OSM"`
* try using an old, pre-.6 version and also run `./martin "postgres://postgres:[pass]@[ip]:5432/OSM"`

**Comment by IceMan1910** [comment]
@nyurik 
> * try running martin without config, i.e. `./martin "postgres://postgres:[pass]@[ip]:5432/OSM"`

same output

> * try using an old, pre-.6 version and also run `./martin "postgres://postgres:[pass]@[ip]:5432/OSM"`

sucessfull with 0.5 version

```
[2022-12-02T16:05:41Z INFO  martin] Starting martin v0.5.0
[2022-12-02T16:05:41Z INFO  martin] Config is not set, scanning database
[2022-12-02T16:05:42Z INFO  martin] Connected to ..://..@...:5432/OSM
[2022-12-02T16:05:42Z INFO  martin::table_source] Found public.boundaries table source
[2022-12-02T16:05:42Z INFO  martin::table_source] Found public.lines table source
[2022-12-02T16:05:42Z INFO  martin::table_source] Found public.points table source
[2022-12-02T16:05:42Z INFO  martin::table_source] Found public.polygons table source
[2022-12-02T16:05:42Z INFO  martin::table_source] Found public.routes table source
[2022-12-02T16:05:44Z INFO  martin] Martin has been started on 0.0.0.0:3000.

```

Is it a bug on 0.6 ?

**Comment by nyurik** [comment]
Hm, this is not good. In 0.6 we switched to a newer postgres accesss lib, async in nature. This (in theory) should not have any consequences. Also, in 0.6 we added tons of itnegration testing on all platforms, including external hosted DBs like digital ocean. All was seemed to be working fine until your report :)  So now we need to figure out how your DB setup or networking is different from all the test setups we have.

So a few things we could try next:
* Try to run another database, perhaps locally using docker (I remapped the port just in case you have another DB running):

```bash
# in shell 1
docker run --rm -it -p 5000:5432 -e POSTGRES_PASSWORD=abc123 postgis/postgis:15-3.3-alpine
# in shell 2
./martin postgres://postgres:abc123@localhost:5000
```

You can try it with both martin versions to see if there are any differences.  I expect it to show a bunch of junk tiger tables discovered as found sources

* See what custom configurations you have done to the database, i.e. any restrictions on who can connect and how? Authentication protocols? SSL? Some other interesting methods?

**Comment by IceMan1910** [comment]
Ok. I will try the docker approach. I'll give feedback asap.

> * See what custom configurations you have done to the database, i.e. any restrictions on who can connect and how? Authentication protocols? SSL? Some other interesting methods?

That's where you are going to be disappointed :\
It's a brand new setup. New Ubuntu 22.0, new postegres 15. From a clean format on the machine. 
Only 1 user on the BD, postegres, only setup done was to open up the standard port to allow for connections. 
0 customization.

**Comment by nyurik** [comment]
@IceMan1910 thx, i wonder if it would be possible to build a docker using those steps?  Basically try to create a `Dockerfile` file that is based on `ubuntu` base image, and the steps would install Postgres and do any needed configuration, plus download martin as last step using curl?  This way we can have a reproducible test.

**Comment by nyurik** [comment]
@IceMan1910 any luck creating a reproducible docker image?  I'm really concerned about this bug. Could it be some old (or new?) authentication method?

**Comment by nyurik** [comment]
@IceMan1910 please confirm you are still having this issue. We need a way to reproduce this bug as it may have a lot of consequences for other users if confirmed.

**Comment by CarlosZiegler** [comment]
@nyurik I am tring to use it:
COMMAND: `docker run --rm -it -p 5001:5432 -e POSTGRES_PASSWORD=abc123 postgis/postgis:15-3.3-alpine`

<details>
<summary>OUTPUT</summary>

```
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... sh: locale: not found
2022-12-12 18:51:00.115 UTC [84] WARNING:  no usable system locales were found
ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....2022-12-12 18:51:01.812 UTC [103] LOG:  starting PostgreSQL 15.1 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r4) 12.2.1 20220924, 64-bit
2022-12-12 18:51:01.814 UTC [103] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2022-12-12 18:51:01.829 UTC [111] LOG:  database system was shut down at 2022-12-12 18:51:01 UTC
2022-12-12 18:51:01.848 UTC [103] LOG:  database system is ready to accept connections
 done
server started

/usr/local/bin/docker-entrypoint.sh: sourcing /docker-entrypoint-initdb.d/10_postgis.sh
CREATE DATABASE
Loading PostGIS extensions into template_postgis
CREATE EXTENSION
CREATE EXTENSION
You are now connected to database "template_postgis" as user "postgres".
CREATE EXTENSION
CREATE EXTENSION
Loading PostGIS extensions into postgres
CREATE EXTENSION
CREATE EXTENSION
You are now connected to database "postgres" as user "postgres".
CREATE EXTENSION
CREATE EXTENSION

waiting for server to shut down...2022-12-12 18:51:10.071 UTC [103] LOG:  received fast shutdown request
.2022-12-12 18:51:10.072 UTC [103] LOG:  aborting any active transactions
2022-12-12 18:51:10.081 UTC [103] LOG:  background worker "logical replication launcher" (PID 117) exited with exit code 1
2022-12-12 18:51:10.083 UTC [107] LOG:  shutting down
2022-12-12 18:51:10.085 UTC [107] LOG:  checkpoint starting: shutdown immediate
2022-12-12 18:51:10.200 UTC [107] LOG:  checkpoint complete: wrote 3803 buffers (23.2%); 0 WAL file(s) added, 0 removed, 2 recycled; write=0.074 s, sync=0.035 s, total=0.116 s; sync files=482, longest=0.014 s, average=0.001 s; distance=31470 kB, estimate=31470 kB
2022-12-12 18:51:10.224 UTC [103] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

2022-12-12 18:51:10.369 UTC [1] LOG:  starting PostgreSQL 15.1 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r4) 12.2.1 20220924, 64-bit
2022-12-12 18:51:10.372 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2022-12-12 18:51:10.372 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2022-12-12 18:51:10.374 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2022-12-12 18:51:10.390 UTC [153] LOG:  database system was shut down at 2022-12-12 18:51:10 UTC
2022-12-12 18:51:10.410 UTC [1] LOG:  database system is ready to accept connections
2022-12-12 18:56:10.447 UTC [149] LOG:  checkpoint starting: time
2022-12-12 18:56:45.896 UTC [149] LOG:  checkpoint complete: wrote 346 buffers (2.1%); 0 WAL file(s) added, 0 removed, 0 recycled; write=35.417 s, sync=0.011 s, total=35.448 s; sync files=103, longest=0.004 s, average=0.001 s; distance=9465 kB, estimate=9465 kB
```

</details>

COMMAND:

```
docker run \
  --net=host \
  -p 3000:3000 \
  -e DATABASE_URL=postgres://postgres:abc123@localhost:5001 \
  maplibre/martin
```

<details>
<summary>OUTPUT</summary>

```
WARNING: Published ports are discarded when using host network mode
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
[2022-12-12T18:51:51Z INFO  martin] Starting Martin v0.6.1
[2022-12-12T18:51:51Z INFO  martin] Config file is not specified
[2022-12-12T18:51:51Z INFO  martin::pg::db] Connecting to database
[2022-12-12T18:51:51Z INFO  martin::pg::db] Automatically detecting table and function sources
[2022-12-12T18:51:52Z INFO  martin::pg::db] No function sources found
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tract.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.addrfeat" table source with "the_geom" column (LINESTRING, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.state" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.bg" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.edges" table source with "the_geom" column (MULTILINESTRING, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.faces.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.bg.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.edges.the_geom" table source with "the_geom" column (MULTILINESTRING, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.faces" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.cousub" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.county" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.cousub.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tabblock" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tabblock.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tabblock20.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.zcta5" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.state.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.addrfeat.the_geom" table source with "the_geom" column (LINESTRING, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.zcta5.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tract" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.place" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.tabblock20" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.place.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin::pg::db] Found "tiger.county.the_geom" table source with "the_geom" column (MULTIPOLYGON, SRID=4269)
[2022-12-12T18:51:52Z INFO  martin] Martin has been started on 0.0.0.0:3000.
[2022-12-12T18:51:52Z INFO  martin] Use http://0.0.0.0:3000/index.json to get the list of available sources.
```

</details>

**Comment by nyurik** [comment]
yes, i'm getting the same result -- seems like everything is working as expected

**Comment by nyurik** [comment]
One important note though -- your docker image seems a bit old. You may want to pull the latest or specify the version:

```
docker run --net host --rm -it maplibre/martin:v0.6.1 postgres://postgres:abc123@localhost:5001
```

P.S.  oops, i misread - you are using the latest 0.6.1

**Comment by carlosrziegler** [comment]
I have the same error right now if try to connect to an external DB ( Google Cloud ). PGAdmin works fine with the same string connection. 
<img width="788" alt="image" src="https://user-images.githubusercontent.com/86771140/209825617-e1f83bca-a9d6-4644-b53f-f647c08b97b3.png">

**Comment by nyurik** [comment]
@carlosrziegler thanks for the feedback. We need a clear way to replicate the issue in order to diagnose it. How do you set up the postgres? What configuration settings do you set on it? What PG version do you use? Ideally, if you can give an exact reproducible steps, especially in a docker container, it would help us immensely. Thanks!

**Comment by nyurik** [comment]
@dwweb0309 I suspect in your case this is due to the martin docker image not having access to the port 5432 from your database's localhost.  Try to run `--net host` instead of `-p 3000:3000`?

**Comment by dwweb0309** [comment]
> @dwweb0309 I suspect in your case this is due to the martin docker image not having access to the port 5432 from your database's localhost. Try to run `--net host` instead of `-p 3000:3000`?

The same result
`docker run --net host -e DATABASE_URL='postgres://postgres:postgres@localhost:5432/propedia' maplibre/martin`

**Comment by nyurik** [comment]
@dwweb0309 are you able to run `psql postgres://postgres:postgres@localhost:5432/propedia`   (from the host)

**Comment by nyurik** [comment]
P.S. one more thing - try to remove the quote symbols around the db url - just in case they get passed incorrectly.  And please describe how your db is set up on the localhost, and what configurations you used. What do `postgresql.conf` and `pg_hba.conf` look like? etc

**Comment by dwweb0309** [comment]
`psql postgres://postgres:postgres@localhost:5432/propedia` works.
My `postgresql.conf` and `pg_hba.conf` are initial state. And I was advised to update `listen_addresses='*'` in `postgresql.conf` and  `host all all 127.0.0.1/32 md5` to `host all all 0.0.0.0/0 md5` in `pg_hba.conf`, but all failed.

**Comment by nyurik** [comment]
@dwweb0309 do you think you could help me debug this issue further? I am currently installing virtualbox and will install ubuntu-server in it, plus install postgres and copy/paste martin binary into it for simplest case.  If that still works ok, will try to set up docker inside the virtualbox.  This way we can have exact reproducible steps to get to the error.

**Comment by nyurik** [comment]
P.S. see if you can get Martin to run without docker (just get the binary for your platform from the releases and run it directly), and if you could do similar steps to get to an error, would be great

**Comment by dwweb0309** [comment]
> @dwweb0309 do you think you could help me debug this issue further? I am currently installing virtualbox and will install ubuntu-server in it, plus install postgres and copy/paste martin binary into it for simplest case. If that still works ok, will try to set up docker inside the virtualbox. This way we can have exact reproducible steps to get to the error.

Thanks!. That is what I did exactly. I am using windows, and have ubuntu 20 in wmware. I have installed postgres 15 in it, and tried to run martin using docker command.

**Comment by nyurik** [comment]
Is there a way you can record every single step you went through? I know its tricky, but it would greatly help us pin down this nasty issue

**Comment by nyurik** [comment]
I might have been able to reproduce it using a downloaded image, but I need to know how you configured postgres.  I used this method:

* install postgres using [these steps](https://www.postgresql.org/download/linux/ubuntu/)
* install postgis with `apt install postgis`
* add a new user + database: `sudo -u postgres createuser -s -P mypg`
* ensure that postgres db has a postgis extension `psql postgres://mypg:mypg@localhost:5432/postgres`
* ensure that it works from a docker: `sudo docker run --rm -it --net host postgres psql postgres://mypg:mypg@localhost:5432/postgres`

**Comment by dwweb0309** [comment]
Thank you for your effort.
Yes, I installed postgres and postgis using [these steps](https://computingforgeeks.com/how-to-install-postgis-on-ubuntu-linux/)
But I didn't add a new user, I am just using **postgres** user with password reset. No more changes in postgres configuration.
Can you try `docker run` with some postgres config changes?

**Comment by ItsCoolOnMars** [comment]
Hello, literally just found Martin 2 hours ago, got similar issues with connecting to the db. Here are my steps:
I've created a new postgres database in Azure with the PostGIS extension. (filled with some test spatial data)
pgAdmin and psql work fine with my connection string/details. 
I tried to test martin from my host machine (Mac). Using Homebrew, it installed v0.5.0 (is that expected?).
I ran martin with the same connection string and got this: `Can't setup connection pool: timed out waiting for connection`.
Found this issue.
And tried to run with docker, got the error message you have here `Can't retrieve connection from the pool: Timed out in bb8` (Docker installed v0.6.2).
I'm not sure how helpful this information is, can provide more details if needed. Really excited to try out the software though!

**Comment by nyurik** [comment]
@ItsCoolOnMars v0.5.0 is extremely old, back from 2019 from even before being part of MapLibre (cc @stepankuzmin) - could you try the latest release please?  We should upload to homebrew.  The more important aspect of your report is that https://github.com/maplibre/martin/issues/496#issuecomment-1335479043 notes that 0.5.0 did work -- which was using different lib to connect to postgres - which makes me think that this might be a different issue.

**Comment by dwweb0309** [comment]
> I might have been able to reproduce it using a downloaded image, but I need to know how you configured postgres. I used this method:
> 
> * install postgres using [these steps](https://www.postgresql.org/download/linux/ubuntu/)
> * install postgis with `apt install postgis`
> * add a new user + database: `sudo -u postgres createuser -s -P mypg`
> * ensure that postgres db has a postgis extension `psql postgres://mypg:mypg@localhost:5432/postgres`
> * ensure that it works from a docker: `sudo docker run --rm -it --net host postgres psql postgres://mypg:mypg@localhost:5432/postgres`

Does above step work? I did the same step, but it doesn't work.
`sudo docker run --rm -it --net host postgresql psql` command sends message like **Unable to find image 'postgres:latest' locally**

**Comment by nyurik** [comment]
@dwweb0309 docker run downloads the `postgres:latest` image automatically - i just tried it on my machine, works fine. If you want, you can first do `docker pull postgres` and then run it.

**Comment by dwweb0309** [comment]
@nyurik 
I have tried to use postgres database in AWS RDS
`sudo docker run --rm -it --net host postgres psql postgres://mypg:mypg@localhost:5432/postgres` works. But `docker run --rm -it -p 3000:3000 -e DATABASE_URL='postgres://rds-user:rds-password@rds-host-url:5432/propedia' maplibre/martin` doesn't work.

**Comment by nyurik** [comment]
@dwweb0309 did you mistype your reply?  The `postgres://...` url should be the same in both docker calls

**Comment by dwweb0309** [comment]
> @dwweb0309 did you mistype your reply? The `postgres://...` url should be the same in both docker calls

`rds-user`, `rds-password`, and `rds-host-url` are just placeholders

**Comment by nyurik** [comment]
thanks, that helps!  I am beginning to suspect that it is the `ssl connection` part that's not working correctly.  When I use psql locally, I do not see that line

**Comment by nyurik** [comment]
does anyone know how to have a simple docker that has a postgresql with ssl, it would be so helpful! esp if it is based on the https://github.com/postgis/docker-postgis -- this way i could try to test it reliably

**Comment by nyurik** [comment]
I think I finally was able to reliably reproduce it.  The issue (suprise!) is related to SSL -- I need to disable hostname verification during SSL handshake.  Secondary issue was upstream -- https://github.com/djc/bb8/issues/151 -- that's why no errors were being reported, instead giving a silly "timeout".

I will fix this in the next few days, and republish, possibly backporting it to 0.6.0 too.

**Comment by cthiange** [comment]
@dwweb0309 in your case it may just be missing a root cert from Amazon. Try making it visible to your container and set the `CA_ROOT_CERT` env variable.

**Comment by nyurik** [comment]
my plan to fix this:
* disable hostname verification during SSL handshake (it will be less secure!), and backport it to 0.6.0
* try to work around the current [limitation](https://github.com/sfackler/rust-postgres/issues/768) of the upstream not supporting all `psql` ssl modes
* migrate bb8 to deadpool, as it seems to be more actively maintained
* migrate to rustls to simplify the build and deployment (remove openssl dependency!)

**Comment by nyurik** [comment]
I just merged the fix to main, could someone with this issue try it?  Note that you would need to either use the `maplibre/martin:main` [docker image](https://hub.docker.com/r/maplibre/martin/tags) or use `cargo run --features ssl --release -- postgres://...` to build it locally.  Thanks!

CC: @IceMan1910 @CarlosZiegler @carlosrziegler @ItsCoolOnMars @dwweb0309

**Comment by ItsCoolOnMars** [comment]
@nyurik I believe you meant `maplibre/martin:main` because "latest" is 2 months old in the docker hub. It didn't work for me with the same error as before. (v0-7-0-alpha)

**Comment by nyurik** [comment]
@ItsCoolOnMars you are absolutely right, thank you, I fixed my comment above.

**Comment by nyurik** [comment]
@ItsCoolOnMars when you said it didn't work -- you meant the latest or the main tag?

**Comment by ItsCoolOnMars** [comment]
well, technically both :). but yeah the "main" one didn't work.

**Comment by dwweb0309** [comment]
> I just merged the fix to main, could someone with this issue try it? Note that you would need to either use the `maplibre/martin:main` [docker image](https://hub.docker.com/r/maplibre/martin/tags) or use `cargo run --features ssl --release -- postgres://...` to build it locally. Thanks!
> 
> CC: @IceMan1910 @CarlosZiegler @carlosrziegler @ItsCoolOnMars @dwweb0309

I tried to use `-c` option, but I always get **No such file or directory** message

**Comment by nyurik** [comment]
I just migrated to dead-pool, so we should never see the timeout error again.  Instead, it will now actually say what the error is.  The docker's `main` image should be published in a few min.  @ItsCoolOnMars could you take a look again and see what error you get now?

@dwweb0309 can you write the whole command please? I'm not certain what you mean.  Also, you linked to the v0.6, and a lot has been changed in the main branch. Thx!

**Comment by ItsCoolOnMars** [comment]
@nyurik With dead-pool it finally works for me: serves the tiles, and everything seems good. Thank you!

**Comment by dwweb0309** [comment]
> I just migrated to dead-pool, so we should never see the timeout error again. Instead, it will now actually say what the error is. The docker's `main` image should be published in a few min. @ItsCoolOnMars could you take a look again and see what error you get now?
> 
> @dwweb0309 can you write the whole command please? I'm not certain what you mean. Also, you linked to the v0.6, and a lot has been changed in the main branch. Thx!

`docker run --rm -it --net=host -p 3009:3000 -e DATABASE_URL=postgres://postgres:password@localhost/propedia maplibre/martin:main -c config.yaml`
`-c 'config.yaml'`, `-c './config.yaml'` all doesn't work

**Comment by dwweb0309** [comment]
@nyurik 
Do you know [tileserver-gl](https://github.com/maptiler/tileserver-gl)? Which is similar to **martin** but it feeds vector tiles from sqlite database. By default, **tileserver-gl** uses config file in the folder where the command runs. I think that is good practice.

**Comment by nyurik** [comment]
@dwweb0309 ah, you forgot to share the file with the docker container itself, and doing the same thing as tileserve-gl wouldn't help here.

try this:

```
docker run --rm -it --net=host -v $PWD:/martin-config -p 3009:3000 \
   -e DATABASE_URL=postgres://postgres:password@localhost/propedia \
   maplibre/martin:main -c /martin-config/config.yaml
```

from the dir that contains your config file (or specify the full path in `-v` param. This way you share the file with the docker container, and it can read it.

**Comment by dwweb0309** [comment]
> @dwweb0309 ah, you forgot to share the file with the docker container itself, and doing the same thing as tileserve-gl wouldn't help here.
> 
> try this:
> 
> ```
> docker run --rm -it --net=host -v $PWD:/martin-config -p 3009:3000 \
>    -e DATABASE_URL=postgres://postgres:password@localhost/propedia \
>    maplibre/martin:main -c /martin-config/config.yaml
> ```
> 
> from the dir that contains your config file (or specify the full path in `-v` param. This way you share the file with the docker container, and it can read it.

I think above comment should be in documentation

**Comment by nyurik** [comment]
Sure! Want to do a pr?

**Comment by nyurik** [comment]
this will be released as part of 0.7.0
